pulp/apis/sas_api.py

```diff
@@ -0,0 +1,623 @@
+                for name, value in values:
+                    f.write(f"{name},{value}\n")
+            return True
+
+        def _get_max_upload_len(self, fileName):
+            maxLen = 0
+            with open(fileName, 'r') as f:
+                for line in f.readlines():
+                    maxLen = max(maxLen, max([len(word) for word in line.split(" ")]))
+            return maxLen + 1
+
+        def actualSolve(self, lp):
+            """Solve a well formulated lp problem"""
+            log.debug("Running SAS")
+
+            if len(lp.sos1) or len(lp.sos2):
+                raise PulpSolverError("SAS94: Currently SAS doesn't support SOS1 and SOS2.")
+
+            postfix = uuid4().hex[:16]
+            tmpMps, tmpMst = self._create_tmp_files(postfix, "mps", "mst")
+            vs = lp.writeMPS(tmpMps, with_objsense=False)
+
+            nameLen = self._get_max_upload_len(tmpMps)
+            if nameLen > MAX_NAME_LENGTH:
+                raise PulpSolverError(f"SAS94: The lengths of the variable or constraint names \
+                                    (including indices) should not exceed {MAX_NAME_LENGTH}.")
+
+            proc = self.proc = "OPTMILP" if self.mip else "OPTLP"
+
+            # Get Obj Sense
+            if lp.sense == constants.LpMaximize:
+                self.solverOptions["objsense"] = "max"
+            elif lp.sense == constants.LpMinimize:
+                self.solverOptions["objsense"] = "min"
+            else:
+                raise PulpSolverError("SAS94 : Objective sense should be min or max.")
+            # Get the rootnode options
+            decomp_str = self._create_statement_str("decomp")
+            decompmaster_str = self._create_statement_str("decompmaster")
+            decompmasterip_str = self._create_statement_str("decompmasterip")
+            decompsubprob_str = self._create_statement_str("decompsubprob")
+            rootnode_str = self._create_statement_str("rootnode")
+
+            if lp.isMIP():
+                if not self.mip:
+                    warnings.warn("SAS94 will solve the relaxed problem of the MILP instance.")
+            # Handle warmstart
+            warmstart_str = ""
+            if self.warmStart:
+                self._write_sol(filename=tmpMst, vs=vs)
+
+                # Set the warmstart basis option
+                if proc == "OPTMILP":
+                    warmstart_str = """
+                                    proc import datafile='{primalin}'
+                                        out=primalin{postfix}
+                                        dbms=csv
+                                        replace;
+                                        getnames=yes;
+                                        run;
+                                    """.format(
+                        primalin=tmpMst,
+                        postfix=postfix,
+                    )
+                    self.solverOptions["primalin"] = f"primalin{postfix}"
```

Comment:
it's not a good practice to edit the `solverOptions` attribute. You should avoid editing during solve. It should be stateless.